Mods to support spectra file IO out of fortran

hbivens opened the pull request.

hbivens committed:
Add install stuff and support for Finding Specutils in CMake

wcjohns committed:
Merge branch 'master' into issue/31-spectra-io-out-of-fortran

wcjohns commented:
This looks like a nice addition - thanks!

hbivens committed:
remove install interface thing that doesn't seem to matter.

hbivens committed:
fortran integration mods

* added separate swig interface file
* build tweaks for successful linking to shared libraries

hbivens committed:
add extension that accesses a count at a given channel

hbivens committed:
make the *_at extensions 1-based, since calling from fortran

hbivens committed:
enable  testing at the topmost cmakefile

* Then you can run ctest in the same dir you build in
* Also update version in unit_tests

hbivens committed:
cherry pick boost test removal

The beauty of doctest is you just have to include one header file.

hbivens committed:
syntax highlighting getting confused

hbivens committed:
Make API a little easier

hbivens committed:
add round trip test

hbivens commented on unit_tests/test_spec_file.cpp:
```diff
@@ -0,0 +1,99 @@
+            //auto timesEqual = expectedM.start_time() == actualM.start_time();
+            auto timeStr1 = SpecUtils::to_iso_string(expectedM.start_time() );
+            auto timeStr2 = SpecUtils::to_iso_string(actualM.start_time() );
+            CHECK( timeStr1 == timeStr2);
```
@wcjohns I think we lose some time precision when we write and read pcf files. My test is failing when comparing measurement start times.

``` 

1: /mnt/c/Projects/code/SpecUtils/unit_tests/test_spec_file.cpp:81: ERROR: CHECK( timeStr1 == timeStr2 ) is NOT correct!
1:   values: CHECK( 20240812T164010.285502 == 20240812T164010.29 )
```

wcjohns replied:
That's correct - PCF files store time in VAX format (a 23 character string formatted like "19-Sep-2014 14:12:01.62"), so only to the 100th of a second.  

SpecUtils normally tracks down to the microsecond (via `using time_point_t = std::chrono::time_point<std::chrono::system_clock,std::chrono::microseconds>`).  

It looks like `SpecUtils::to_vax_string(..)` rounds to the nearest 0.01 seconds, so you could test:
```
auto timeStr1 = SpecUtils::to_vax_string( expectedM.start_time() );
auto timeStr2 = SpecUtils::to_vax_string( actualM.start_time() );
CHECK( timeStr1 == timeStr2);
```

Or something like:

```
CHECK( std::round(0.0001*expectedM.start_time().count()) == std::round(0.0001*actualM.start_time().count()) );
```

hbivens replied:
Ok, thanks! I'll use `to_vax_string` to compare.

hbivens committed:
use seconds precision

hbivens committed:
add deviation pairs to test

hbivens committed:
check neutrons

hbivens committed:
use vax string

hbivens committed:
Yeah!!! Figured out fortran array to c++ typemap.

hbivens committed:
add a few more extensions and include filesystem stuff

hbivens committed:
call correct method

hbivens committed:
trim the filename and add some error diagnostics

hbivens committed:
add realtime check

hbivens committed:
merge master

wcjohns committed:
Fix issue compiling with fast_float

wcjohns commented:
Commit 6d8371f39d2eef24edb123a25abc738de43b13a3 *should* fix this - let me know if not.

hbivens committed:
fix windows build error

hbivens committed:
update wrapper

hbivens committed:
start some tests to map dev pairs

hbivens committed:
make strtod default

hbivens committed:
Merge remote-tracking branch 'origin/master' into issue/31-spectra-io-out-of-fortran

wcjohns commented on CMakeLists.txt:
If `SpecUtils_FLT_PARSE_METHOD` is left as 'default_value' here, there is some logic below to hopefully pick the most reasonable float parsing method.  Like for MSVC &ge;2019, it *should* pick 'FromChars'.
Is there is  reason to default to the slowest common denominator, or was it giving you an issue?

hbivens commented:
The default_value wasn't working for me, and I think that was before you fixed the FastFloat retrieval. I can revert it back to default_value.

-Hugh

____________________________________________________
Hugh P. Bivens
Sandia National Laboratories
Albuquerque, NM USA
+1 (505) 284-6822
https://snl-wiki.sandia.gov/x/6ID2BQ

From: William Johnson ***@***.***>
Sent: Monday, September 16, 2024 5:43 PM
To: sandialabs/SpecUtils ***@***.***>
Cc: Bivens, Hugh P. ***@***.***>; Author ***@***.***>
Subject: [EXTERNAL] Re: [sandialabs/SpecUtils] Mods to support spectra file IO out of fortran (PR #34)


@wcjohns commented on this pull request.

________________________________

On CMakeLists.txt<https://github.com/sandialabs/SpecUtils/pull/34#discussion_r1762087752>:

If SpecUtils_FLT_PARSE_METHOD is left as 'default_value' here, there is some logic below to hopefully pick the most reasonable float parsing method. Like for MSVC ≥2019, it should pick 'FromChars'.
Is there is reason to default to the slowest common denominator, or was it giving you an issue?

-
Reply to this email directly, view it on GitHub<https://github.com/sandialabs/SpecUtils/pull/34#pullrequestreview-2308000642>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADAP4BE7WQ6LSBY52EADLQLZW5UJTAVCNFSM6AAAAABLPPWDBOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGMBYGAYDANRUGI>.
You are receiving this because you authored the thread.Message ID: ***@***.******@***.***>>

hbivens committed:
checkpoint

hbivens committed:
pcf spectra file mods

hbivens committed:
updates - add fortran round trip test

hbivens committed:
checkpoint

* add pcf extension classes - this is the way

hbivens committed:
lose the pcf namespace

hbivens committed:
figured out deviation pairs

hbivens committed:
update fortran wrapper

hbivens committed:
more dev pair mapping

hbivens committed:
cherrypick Will's tag support

hbivens committed:
merge master and fix conflicts

hbivens committed:
revert back to default

hbivens committed:
update test to show Will

wcjohns committed:
Fix issues with writing and reading deviation pairs to PCF files.

The ordering of deviation pairs was not being handled consistently across writing and reading pcf files.  Cleaned this all up a bit, and documented things a bit better.

wcjohns commented on unit_tests/test_spec_file.cpp:
```diff
@@ -156,8 +156,8 @@ TEST_CASE("Round Trip")
         //auto detNames = generateDetectorNames();
-        //std::vector<std::string> detNames = { "Ba1", "Aa2", "Bc3", "Cb4" }; // Bc3 computes an out of range index
-        std::vector<std::string> detNames = { "Ba1", "Aa2", "Bb3", "Cb4" };
+        std::vector<std::string> detNames = { "Ba1", "Aa2", "Bc3", "Cb4" }; // Bc3 computes an out of range index
```
Commit 431c65ce65860c2421022b26b0f385da8c2a78f8 fixes this issue.  
Bc3 is a valid name - the indexing of deviation pairs inside the PCF array for them, was not being handled consistently when reading and writing PCF files.
Thanks for finding this!

wcjohns committed:
Make compiling the bindings optional (defaulting to off).

Add the CI to use/test the C bindings, and for linux, the Fortran.

wcjohns committed:
Promote measurement source and descriptions to be member variables of Measurement.

Previously was storing and accessing this information in the `remarks_` fields, with prefixes of "Description:" and "Source:".
Full support of all PCF fields was the impetus for these changes, but there were a few other places these fields were effectively used as well.

It would be nice to properly support this information by altering it to conform to N42-2012 <RadItemInformation> elements.

wcjohns committed:
Remove PcfExtensions.

These additions have been integrated into the primary `SpecFile` and `Measurement` classes.

wcjohns committed:
Fix compile issue.

Not sure if PcfUtils.h is supposed to exist?

wcjohns committed:
Fix some function not always returning values in c-interface.

wcjohns committed:
Fix `SpecUtils_Measurement_set_start_time_str` not returning a value.

hbivens committed:
update fortran bindings

hbivens committed:
add ctest args

hbivens committed:
rename this, I was getting compiler warning:

/mnt/c/Projects/code/SpecUtils/unit_tests/test_spec_file.cpp: In function ‘void DOCTEST_ANON_FUNC_14()’:
/mnt/c/Projects/code/SpecUtils/unit_tests/test_spec_file.cpp:180:55: warning: ISO C++ says that these are ambiguous, even though the worst conversion for the first is better than the worst conversion for the second:
  180 |             auto &m = *(specfile.measurement(size_t(0)) );
      |                                                       ^
In file included from /mnt/c/Projects/code/SpecUtils/unit_tests/test_spec_file.cpp:5:
/mnt/c/Projects/code/SpecUtils/SpecUtils/SpecFile.h:1460:38: note: candidate 1: ‘std::shared_ptr<const SpecUtils::Measurement> SpecUtils::SpecFile::measurement(size_t) const’
 1460 |   std::shared_ptr<const Measurement> measurement( size_t num ) const;
      |                                      ^~~~~~~~~~~
/mnt/c/Projects/code/SpecUtils/SpecUtils/SpecFile.h:2447:32: note: candidate 2: ‘std::shared_ptr<SpecUtils::Measurement> SpecUtils::SpecFile::measurement(std::shared_ptr<const SpecUtils::Measurement>)’
 2447 |   std::shared_ptr<Measurement> measurement( std::shared_ptr<const Measurement> meas );

hbivens committed:
add debug print for CICD

hbivens committed:
mas debug

hbivens committed:
add index, size to exception message

hbivens committed:
add size

hbivens committed:
add build script to aid in cicd debug

hbivens committed:
up minimum to avoid cmake warning

hbivens committed:
uri spectra causes fortran test to fail

hbivens committed:
wrong line

hbivens commented on .github/workflows/ci.yml:
```diff
@@ -43,31 +43,31 @@ jobs:
     - name: Configure CMake on Linux
       if: runner.os == 'Linux'
-      run: cmake -S ${{ github.workspace }} -B build -DSpecUtils_BUILD_UNIT_TESTS=ON -DSpecUtils_BUILD_REGRESSION_TEST=ON -DSpecUtils_ENABLE_EQUALITY_CHECKS=ON -DPERFORM_DEVELOPER_CHECKS=ON -DSpecUtils_ENABLE_D3_CHART=ON -DSpecUtils_D3_SUPPORT_FILE_STATIC=ON -DSpecUtils_ENABLE_URI_SPECTRA=ON -DCMAKE_BUILD_TYPE=Release -DSpecUtils_FLT_PARSE_METHOD=${{ matrix.num-parse-method }} -DSpecUtils_FETCH_F ...
+      run: cmake -S ${{ github.workspace }} -B build -DSpecUtils_BUILD_UNIT_TESTS=ON -DSpecUtils_BUILD_REGRESSION_TEST=ON -DSpecUtils_ENABLE_EQUALITY_CHECKS=ON -DPERFORM_DEVELOPER_CHECKS=ON -DSpecUtils_ENABLE_D3_CHART=ON -DSpecUtils_D3_SUPPORT_FILE_STATIC=ON -DSpecUtils_ENABLE_URI_SPECTRA=OFF -DCMAKE_BUILD_TYPE=Release -DSpecUtils_FLT_PARSE_METHOD=${{ matrix.num-parse-method }} -DSpecUtils_FETCH_ ...
```
@wcjohns I'm not sure why but URI Spectra breaks my fortran unit test. When I read in the file I write in the round trip test then specFile doesn't have any measurements.

wcjohns committed:
First try at adding an automated release of shared libraries for windows

wcjohns committed:
Merge branch 'issue/31-spectra-io-out-of-fortran' of github.com:sandialabs/SpecUtils into issue/31-spectra-io-out-of-fortran

wcjohns committed:
Add package command to CMake, and bring out option for Windows runtime, even if not python et al

wcjohns committed:
Merge branch 'issue/31-spectra-io-out-of-fortran' of github.com:sandialabs/SpecUtils into issue/31-spectra-io-out-of-fortran

wcjohns committed:
Merge branch 'master' into issue/31-spectra-io-out-of-fortran

wcjohns committed:
Fix a few things up that I think got messed up during various merges.

wcjohns committed:
Maybe fix compile error on Windows

wcjohns committed:
Add another missing include

wcjohns committed:
Add another missing include

wcjohns committed:
Ad UTF-8 BOM to test_utf8_limit_str_size.cpp

wcjohns committed:
Update ci.yml to upload Windows artifact, maybe

wcjohns committed:
Update ci.yml to upload Windows artifact, now a little closer maybe

wcjohns committed:
Update ci.yml to upload Windows artifact, try again

wcjohns committed:
Update ci.yml to upload Windows artifact, try again

wcjohns committed:
Update ci.yml to upload Windows artifact, try again

wcjohns committed:
Update ci.yml to upload Windows artifact, try again

wcjohns committed:
Minor change

wcjohns committed:
Fix issue when converting to a VAX string.

Before if time rounded to 100 hundreths of a second, that would be printed, instead of incrementing things properly.
For example "2014-Sep-19 23:59:59.996" would print "19-Sep-2014 23:59:59.100", when it should print "20-Sep-2014 00:00:00.00" - which it now does.

wcjohns committed:
Debug windows automated build of shared libraries a bit

wcjohns committed:
Maybe fix syntax error in ci.yml

wcjohns committed:
Maybe fix syntax error in ci.yml

wcjohns committed:
Maybe fix syntax error in ci.yml

wcjohns committed:
Try to improve CI install targets

wcjohns committed:
Debugging chany ci.yml

wcjohns committed:
Remove trying to install export file

wcjohns committed:
Turn building C bindings on for CI build

wcjohns committed:
Add exporting classes/structs/functions for Windows DLLs.

wcjohns committed:
Trye having CMake generate export header.

wcjohns committed:
Fix missed export

wcjohns committed:
Temporarily re-include `pcf_det_name_to_dev_pair_index` so fortran bindings will build.

wcjohns committed:
Fix infinite recursion

hbivens committed:
add script

hbivens committed:
Merge remote-tracking branch 'origin/issue/31-spectra-io-out-of-fortran' into issue/31-spectra-io-out-of-fortran

hbivens committed:
add install

hbivens committed:
Turn off c-bindings

hbivens committed:
add -fPic

hbivens committed:
Revert "Turn building C bindings on for CI build"

This reverts commit 53b503986377d0fb1d6ecd3e3e1fe4e85a6973b8.

hbivens committed:
Revert "Add exporting classes/structs/functions for Windows DLLs."

This reverts commit 160b5bd2692b903dd9fade0dd0b6243685f3c0d2.

hbivens committed:
update swig-gen

hbivens committed:
Revert "Fix missed export"

This reverts commit 13592289193ddb89f007663c994cd3a63cfb0df3.

hbivens committed:
make it work

hbivens committed:
keep the swig generated files in repo

add option to turn on swig fortran wrapper generation

hbivens committed:
update interface

hbivens committed:
update test

hbivens committed:
Add method to update detector name

get mutable reference ecal and dev pairs

hbivens committed:
cmake mods to enable valgrind memcheck with ctest

hbivens committed:
update version and add to cpack file

* error diags to cubic spline

hbivens committed:
Merge remote-tracking branch 'origin/master' into issue/31-spectra-io-out-of-fortran

hbivens committed:
Update SpecUtils_config.h.in

hbivens committed:
maybe fix macos wheel build

hbivens commented on CMakeLists.txt:
```diff
@@ -11,6 +11,9 @@ endif()
 enable_testing()  # Enable here to allow running `ctest` from top-most build dir
 
+# Add this line to specify the architectures for MacOS
+set(CMAKE_OSX_ARCHITECTURES "x86_64;arm64" CACHE STRING "Build architectures for MacOS" FORCE)
```
@wcjohns so everything builds ok but macos*. Copilot said to add this but that didn't help.

hbivens replied:
there's `Explain Error` button:

![image](https://github.com/user-attachments/assets/2fb7f9b0-0cd7-41aa-b3c0-c58373d44cd6)

wcjohns replied:
We can probably remove that like; I think the macOS architectures are set by the nanobind stuff.

I just triggered a build on the main branch, and it ran fine.  
It looks like the problem is that on this branch, it is installing all the headers, and I *think* crucially the static SpecUtils.a library, into the Python wheel.  When `delocate` is ran on the wheel, the .a file is causing the failure.

Below is the step that is installing stuff into the python package:

![image](https://github.com/user-attachments/assets/a55e5484-c0b4-4cb1-a78e-57823de454ef)
![image](https://github.com/user-attachments/assets/688c9184-2f77-4745-8b9b-5544b94df5a1)


And the actual error:
![image](https://github.com/user-attachments/assets/6ecc5b23-e2a3-46d8-9afc-cd8977ecb4a2)
Which looks to be being triggered by the .a file.

hbivens commented:
You're saying that installing the headers is causing it to break?

____________________________________________________
Hugh P. Bivens
Sandia National Laboratories
Albuquerque, NM USA
+1 (505) 284-6822
https://snl-wiki.sandia.gov/x/6ID2BQ

From: William Johnson ***@***.***>
Sent: Wednesday, January 29, 2025 5:27 PM
To: sandialabs/SpecUtils ***@***.***>
Cc: Bivens, Hugh P. ***@***.***>; Author ***@***.***>
Subject: [EXTERNAL] Re: [sandialabs/SpecUtils] Mods to support spectra file IO out of fortran (PR #34)


@wcjohns commented on this pull request.

________________________________

In CMakeLists.txt<https://github.com/sandialabs/SpecUtils/pull/34#discussion_r1934851751>:

> @@ -11,6 +11,9 @@ endif()

 project( SpecUtils VERSION 2025.1.0 )

 enable_testing()  # Enable here to allow running `ctest` from top-most build dir



+# Add this line to specify the architectures for MacOS

+set(CMAKE_OSX_ARCHITECTURES "x86_64;arm64" CACHE STRING "Build architectures for MacOS" FORCE)

We can probably remove that added line; I think the macOS architectures are set by the nanobind stuff.

I just triggered a build on the main branch, and it ran fine.
It looks like the problem is that on this branch, it is installing all the headers, and I think crucially the static SpecUtils.a library, into the Python wheel. When delocate is ran on the wheel, the .a file is causing the failure.

Below is the step that is installing stuff into the python package:

image.png (view on web)<https://github.com/user-attachments/assets/a55e5484-c0b4-4cb1-a78e-57823de454ef>
image.png (view on web)<https://github.com/user-attachments/assets/688c9184-2f77-4745-8b9b-5544b94df5a1>

And the actual error:
image.png (view on web)<https://github.com/user-attachments/assets/6ecc5b23-e2a3-46d8-9afc-cd8977ecb4a2>
Which looks to be being triggered by the .a file.

-
Reply to this email directly, view it on GitHub<https://github.com/sandialabs/SpecUtils/pull/34#discussion_r1934851751>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADAP4BEL3FBAP6FG6EFTHJT2NFWWRAVCNFSM6AAAAABLPPWDBOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKOBSGY3DQNBXHA>.
You are receiving this because you authored the thread.Message ID: ***@***.******@***.***>>

wcjohns committed:
Make so install stuff is only called if building as a top-level project.

This is in respect to trying to fix macOS CI build: https://github.com/sandialabs/SpecUtils/pull/34#discussion_r1934851751

wcjohns committed:
Fix last commit not generating SpecUtils_config.h

wcjohns commented:
It *looks* like its libSpecUtils.a that is causing the issue; I'm not sure why delocate is trying to do its thing on a static library, I thought it essentially fixed up shared libraries.

I just pushed 4e708281757837b820658d3d17aa2209fe760b2c and 4f0f40253d08ff5de52aef91aaf93533e478eff3, which changes the main CMakeLists.txt file so none of the install stuff is called, if the build is not a top-level project.  E.g., if a python build.  
It *seems* like this is probably the behavior wanted, in general - does it seem like it to you?

hbivens commented:
> It _looks_ like its libSpecUtils.a that is causing the issue; I'm not sure why delocate is trying to do its thing on a static library, I thought it essentially fixed up shared libraries.
> 
> I just pushed [4e70828](https://github.com/sandialabs/SpecUtils/commit/4e708281757837b820658d3d17aa2209fe760b2c) and [4f0f402](https://github.com/sandialabs/SpecUtils/commit/4f0f40253d08ff5de52aef91aaf93533e478eff3), which changes the main CMakeLists.txt file so none of the install stuff is called, if the build is not a top-level project. E.g., if a python build. It _seems_ like this is probably the behavior wanted, in general - does it seem like it to you?

Yeah I don't think we need to install stuff on a python build.

hbivens committed:
allow json in repo

hbivens committed:
add presets

hbivens committed:
add some helpful descriptions

hbivens commented:
@wcjohns How do I set or get the energy cal low energy field?

![image](https://github.com/user-attachments/assets/d6618af1-39f7-4cf4-a5cc-bfaf1833d7d7)

wcjohns commented:
Its the 5th entry in the energy calibration coefficient parameters.
For example, its used [here](https://github.com/sandialabs/SpecUtils/blob/0b9939e3aa6348c88558df678482d449f6cfe9e7/src/EnergyCalibration.cpp#L802).

hbivens commented:
> Its the 5th entry in the energy calibration coefficient parameters. For example, its used [here](https://github.com/sandialabs/SpecUtils/blob/0b9939e3aa6348c88558df678482d449f6cfe9e7/src/EnergyCalibration.cpp#L802).

Thanks!

hbivens committed:
need chrono

hbivens committed:
updates

hbivens committed:
updates

hbivens committed:
save

hbivens committed:
cmake tweaks, update version

hbivens committed:
Merge branch 'fix-branch' into issue/31-spectra-io-out-of-fortran

hbivens committed:
merge remote - I guess

hbivens committed:
just name it SpecUtils cuz it's easier :-)

hbivens committed:
allow spec-file overwrite

hbivens committed:
initial izae to zero and diable some errors

hbivens committed:
allow overwriting files

hbivens committed:
maybe allow writing empty pcf files

hbivens committed:
build mods

hbivens committed:
remove comment

hbivens committed:
more swig goodness :-)

hbivens committed:
return reference

hbivens committed:
add more stuff

hbivens committed:
add win32 check

hbivens committed:
fix typo

hbivens committed:
rev version

hbivens committed:
parallelism and add tests

hbivens committed:
swig regen
